### ISSUE TITLE
Invalid testnet block

### ISSUE THREAD

**Issue by jameshilliard** [bitprim/bitprim-mining#1]
It seems this software mined [this invalid block](https://live.blockcypher.com/btc-testnet/block/000000000000034ba26cee29d86f20d6ead376aa1b868c14fd13bacac54959de/) on testnet, the coinbase tag was “Bitprim Project”.

> morcos: sdaftuar took a quick look earlier and said it looked like there was a witness commitment in coinbase but at least one tx that needed a witness was missing it .

**Comment by fivepiece** [comment]
And a trace of what happened:

https://gist.github.com/fivepiece/c6941891694bdf4f64d2a49ecae80853

**Comment by jameshilliard** [comment]
Apparently this software doesn't actually support segwit, even though it produces templates that indicates it does.

**Comment by fivepiece** [comment]
@jameshilliard I guess you're right!
https://github.com/bitprim/bitprim-mining/blob/master/include/bitcoin/mining/getblocktemplate.hpp#L552

**Comment by jameshilliard** [comment]
@fivepiece Yep, no idea why it would flag for segwit witness serialization there if not actually supported. They switched from ckpool to NOMP apparently because they misdiagnosed the issue as being with ckpool, ckpool was serializing the witness as requested(which they should not have been requesting of course). NOMP however would work fine with that because it doesn't support internal witness serialization and instead just uses the GBT default witness commitment if present(it ignores the rules basically).

**Comment by fivepiece** [comment]
Still what's really weird, is that the (wrong) witness commitment is found in the coinbase transaction, so _something_ was working on the witness part.  They got the wtxid of the coinbase tx correct (easy, 0x0000..), so it's almost "half-complete" in that regard.
Why even add the standard witness op_return output if you're actively trying to avoid segwit?  I could understand if it was "only" treating the segwit transaction as some non-standard and mining it, but adding the witness commitment is weird.

**Comment by jameshilliard** [comment]
The stratum server inserts the commitment OP_RETURN, so that would be ckpool doing that in response to the GBT rules containing the segwit rule.

**Comment by fivepiece** [comment]
@jameshilliard , ah!  Understood, was missing that part first.
All makes sense now.
